### PR TITLE
SQL: Make dangerous functions more granular

### PR DESCRIPTION
**Pull request by jkosh44**
Previously, all functions within the mz_internal schema were declared
dangerous, and we disallowed users from executing any of them. This was too broad and prevented users from executing useful, but not stabilized functions, such as is_rbac_enabled.

This commit adds a new system schema, mz_unsafe, that contains all the dangerous functions. Users can execute functions from mz_internal, but not mz_unsafe.

### Motivation
This PR adds a feature that has not yet been specified.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Allow users to call safe but unstable functions (such as `is_rbac_enabled`).

### CODE REVIEW

**Comment by benesch** [comment]
What do you think about calling the schema `mz_unsafe`, to match the `--unsafe-mode` flag and `unsafe_*` feature flags?

**Comment by jkosh44** [comment]
> What do you think about calling the schema `mz_unsafe`, to match the `--unsafe-mode` flag and `unsafe_*` feature flags?

I like that idea, I'll push an update.

**Comment by jkosh44** [comment]
> > What do you think about calling the schema `mz_unsafe`, to match the `--unsafe-mode` flag and `unsafe_*` feature flags?
> 
> I like that idea, I'll push an update.

Though, just so you're aware, the feature flag for this specific feature is `enable_dangerous_functions`.

**Comment by benesch** [comment]
> enable_dangerous_functions

Gah! Can we change it to `enable_unsafe_functions`?

**Comment by jkosh44** [comment]
> > enable_dangerous_functions
> 
> Gah! Can we change it to `enable_unsafe_functions`?

https://github.com/MaterializeInc/materialize/pull/23665

Hopefully it doesn't break any of the upgrade tests.

**Review by benesch — APPROVED** [review]
This is a great idea. I'm signed off on the idea. I didn't look at the implementation though.

**Comment by jkosh44** [comment]
Note, a previous approach that adds a flag to all functions can be found here: https://github.com/MaterializeInc/materialize/pull/22872. I like this approach better, but in case others are curious they can take a look.

File: test/incident-72/verify-heal-in-memory.td [hunk 1]
```diff
@@ -43,6 +43,6 @@ $ set-from-sql var=source-id
 SELECT id FROM mz_objects WHERE name = 'textavro';
 
 $ set-from-sql var=source-shard-id
-SELECT shard_id FROM mz_internal.mz_storage_shards WHERE object_id = '${source-id}';
+SELECT shard_id FROM mz_unsafe.mz_storage_shards WHERE object_id = '${source-id}';
```
Comment (ParkMyCar) [inline test/incident-72/verify-heal-in-memory.td:46]: It seems like this shouldn't have changed?

**Review by ParkMyCar — APPROVED** [review]
Looks good to me!

Just a thought, for a future PR, what if we named the schemas the following:

* `mz_internal`: Functions that should never be called by external users, e.g. `mz_sleep`, i.e. what `mz_unsafe` is in this PR
* `mz_unstable`: Functions that are unstable, but callable by users, i.e. what `mz_internal` is in this PR

File: test/incident-72/verify-heal-in-memory.td (hunk 1 above)
Comment (jkosh44) [reply test/incident-72/verify-heal-in-memory.td]: Yep, fixed.

**Comment by jkosh44** [comment]
> Looks good to me!
> 
> Just a thought, for a future PR, what if we named the schemas the following:
> 
>   * `mz_internal`: Functions that should never be called by external users, e.g. `mz_sleep`, i.e. what `mz_unsafe` is in this PR
> 
>   * `mz_unstable`: Functions that are unstable, but callable by users, i.e. what `mz_internal` is in this PR

I think this is a great idea! I'll open an issue.

**Comment by jkosh44** [comment]
@MaterializeInc/testing The upgrade tests are failing because `mz_unsafe` is unknown in previous versions. Any thoughts on how to fix that?

**Pull request merged**